src/battle_util.c

```diff
@@ -9122,11 +9129,16 @@ bool32 CanMegaEvolve(u8 battlerId)
 
     // Check if trainer already mega evolved a pokemon.
     if (mega->alreadyEvolved[battlerPosition])
+    // should be: if (gBattleStruct->mega.alreadyEvolved[battlerPosition])
+    
         return FALSE;
     if (gBattleTypeFlags & BATTLE_TYPE_DOUBLE)
     {
         if (IsPartnerMonFromSameTrainer(battlerId)
+            
             && (mega->alreadyEvolved[partnerPosition] || (mega->toEvolve & gBitTable[BATTLE_PARTNER(battlerId)])))
+            // should be: && (gBattleStruct->mega.alreadyEvolved[partnerPosition] || (gBattleStruct->mega.toEvolve & gBitTable[BATTLE_PARTNER(battlerId)])))
+            
             return FALSE;
```

Comment:
The suggested change in this file until this point should be done in my opinion, but I wouldn't mind someone taking a good look at them.

From here the changes are just better because it fixes a bug with the fervent wish string.